src/metpy/io/gempak.py

```diff
@@ -0,0 +1,2830 @@
+            fmt = (fmt[0], self.prefmt + fmt[1] if fmt[1] != 's' else fmt[1])
+            for n, part in enumerate(self.parts):
+                for _ in range(part.parameter_count):
+                    if 's' in fmt[1]:
+                        self.parameters[n][attr] += [
+                            self._buffer.read_binary(*fmt)[0].decode().strip()
+                        ]
+                    else:
+                        self.parameters[n][attr] += self._buffer.read_binary(*fmt)
+
+    def _swap_bytes(self, binary):
+        """Swap between little and big endian."""
+        self.swaped_bytes = (struct.pack('@i', 1) != binary)
+
+        if self.swaped_bytes:
+            if sys.byteorder == 'little':
+                self.prefmt = '>'
+                self.endian = 'big'
+            elif sys.byteorder == 'big':
+                self.prefmt = '<'
+                self.endian = 'little'
+        else:
+            self.prefmt = ''
+            self.endian = sys.byteorder
+
+    def _process_gempak_header(self):
+        """Read the GEMPAK header from the file."""
+        fmt = [('text', '28s', bytes.decode), (None, None)]
+
+        header = self._buffer.read_struct(NamedStruct(fmt, '', 'GempakHeader'))
+        if header.text != GEMPAK_HEADER:
+            raise TypeError('Unknown file format or invalid GEMPAK file')
+
+    @staticmethod
+    def _convert_dattim(dattim):
+        """Convert GEMPAK DATTIM integer to datetime object."""
+        if dattim:
+            if dattim < 100000000:
+                dt = datetime.strptime(str(dattim), '%y%m%d')
+            else:
+                dt = datetime.strptime('{:010d}'.format(dattim), '%m%d%y%H%M')
+        else:
+            dt = None
+        return dt
+
+    @staticmethod
+    def _convert_ftime(ftime):
+        """Convert GEMPAK forecast time and type integer."""
+        if ftime >= 0:
+            iftype = ForecastType(ftime // 100000)
+            iftime = ftime - iftype.value * 100000
+            hours = iftime // 100
+            minutes = iftime - hours * 100
+            out = (iftype.name, timedelta(hours=hours, minutes=minutes))
+        else:
+            out = None
+        return out
+
+    @staticmethod
+    def _convert_level(level):
+        """Convert levels."""
+        if isinstance(level, (int, float)) and level >= 0:
+            return level
+        else:
+            return None
+
+    @staticmethod
+    def _convert_vertical_coord(coord):
+        """Convert integer vertical coordinate to name."""
+        if coord <= 8:
+            return VerticalCoordinates(coord).name.upper()
+        else:
+            return struct.pack('i', coord).decode()
+
+    @staticmethod
+    def _convert_parms(parm):
+        """Convert parameter strings."""
+        dparm = parm.decode()
+        return dparm.strip() if dparm.strip() else ''
+
+    @staticmethod
+    def _fortran_ishift(i, shift):
+        """Python-friendly bit shifting."""
+        mask = 0xffffffff
```

Comment:
Because of how Python [handles negative numbers](https://wiki.python.org/moin/BitwiseOperators), I had to implement it the way I did to ensure the correct number of bits was in each integer. My testing seems to indicate I can drop the `ctypes` stuff if I properly apply the bitmask. I do not think I can get it any simpler than that without improperly unpacking data.

Reply:
Ah yes, Python's signed integers and shifting. I think *this* should do it (took way to much to figure out):
```python
def my_shift(i, shift):
    if shift >= 0:
        # Shift left and only keep low 32 bits
        ret = (i << shift) & 0xffffffff
        
        # If high bit, convert back to negative of two's complement
        if ret > 0x7fffffff:
            ret = -(~ret & 0x7fffffff) - 1
        return ret
    elif shift < 0:
        # Shift right the low 32 bits
        return (i & 0xffffffff) >> -shift
```